### PR TITLE
pkgdir: new check for total FILESDIR size

### PR DESCRIPTION
**Pull request by mgorny**
Moves FILESDIR file limit from hardcoded values to class var, and adds a check for total FILESDIR size.

### CODE REVIEW

**Review by arthurzam — APPROVED** [review]
Thanks

File: src/pkgcheck/checks/pkgdir.py
Comment (mgorny) [inline src/pkgcheck/checks/pkgdir.py:282]: FYI, I was considering something along the lines of:

```suggestion
        if sum(size*len(files) for size, files in files_by_size.items()) > TotalSizeViolation.limit:
```

but I've decided that plain `+=` should be both more readable and faster.

**Comment by mgorny** [comment]
Ok, I've added a check for Manifest size as well. Put a random figure of 100 KiB per package version for now.

I've made the limit dependent on number of unique package versions, so that e.g. version bumps don't cause packages to suddenly just over limit, and then cleanups put them below again. So e.g. a package with 3 versions get 300 KiB limit.

With this assumptions (i.e. size / number of versions), the top Manifests in ::gentoo are:

<details><summary>Biggest Manifests in Gentoo (per ebuild version)</summary>

```
1095871	dev-texlive/texlive-latexextra-2021
699709	app-metrics/prometheus-2.31.1
305354	games-board/tablebase-syzygy-0_pre20131213
298904	net-fs/minio-2021.04.18.19.26.29-r2
269967	app-misc/spire-1.1.0
266770	net-fs/mc-2021.03.23.05.46.11
226363	dev-texlive/texlive-fontsextra-2021
207623	app-admin/filebeat-7.16.2
184518	dev-texlive/texlive-publishers-2021
168333	dev-texlive/texlive-mathscience-2021
149972	dev-texlive/texlive-pictures-2021
147704	app-metrics/memcached_exporter-0.9.0
147618	app-metrics/postgres_exporter-0.9.0
146924	net-analyzer/chronograf-1.8.5
137762	app-containers/distrobuilder-1.1
121527	net-misc/rclone-1.53.3
116566	app-crypt/sequoia-sq-0.26.0
102704	dev-texlive/texlive-bibtexextra-2021
95816	dev-vcs/reposurgeon-4.27
78972	dev-texlive/texlive-pstricks-2021
77948	app-admin/gopass-jsonapi-1.11.1
76738	app-metrics/node_exporter-1.1.2-r1
76720	app-metrics/blackbox_exporter-0.18.0-r2
76291	app-backup/restic-0.11.0
70861	net-analyzer/telegraf-1.13.3
68941	app-admin/gopass-hibp-1.12.0
68414	dev-texlive/texlive-plaingeneric-2021
67037	dev-go/golicense-0.2.0
66250	app-text/mdbook-0.4.18
63381	net-p2p/arti-0.3.0
62840	app-shells/nushell-0.43.0
62805	app-admin/git-credential-gopass-1.12.0
62618	app-office/libreoffice-l10n-7.2.6.2
61486	dev-embedded/arduino-builder-1.3.25
60053	app-text/texlive-core-2021-r1
59128	dev-texlive/texlive-latexrecommended-2021
58767	dev-util/sccache-0.2.15
57087	dev-go/delve-1.7.2
56431	app-i18n/yaskkserv2-0.1.1-r1
54971	sys-process/below-0.5.0
52456	dev-util/cargo-audit-0.15.0
51804	app-admin/gopass-1.12.6
50009	dev-util/wasmer-0.11.0
49453	sys-apps/ripgrep-all-0.9.6-r1
49254	dev-texlive/texlive-latex-2021
48267	sys-apps/uutils-0.0.12
45983	app-crypt/sequoia-sqv-1.1.0
45476	app-admin/gopass-summon-provider-1.12.0
44497	net-vpn/vpncloud-2.3.0
44434	dev-texlive/texlive-langeuropean-2021
43783	app-metrics/fritzbox_smarthome_exporter-0.0.8
43675	dev-texlive/texlive-luatex-2021
42686	app-emulation/ruffle-0_p20220516
41890	dev-go/go-license-detector-3.1.0
41611	dev-texlive/texlive-humanities-2021
41085	app-text/tessdata_legacy-4.0.0
41060	dev-texlive/texlive-langother-2021
40385	dev-util/rustup-1.24.3
40044	app-text/tessdata_best-4.0.0
39979	app-text/tessdata_fast-4.0.0
39923	dev-texlive/texlive-langenglish-2021
39622	dev-db/influxdb-1.8.0
38978	dev-texlive/texlive-langjapanese-2021
38253	net-misc/websocat-1.6.0-r2
36987	media-video/rav1e-0.4.1
35028	app-misc/weggli-0.2.3
34965	dev-util/maturin-0.12.17
34602	dev-python/nltk-data-20211221
33503	dev-util/kubebuilder-2.3.1
32881	dev-util/tree-sitter-cli-0.20.6
32084	www-apps/nextcloud-notify_push-0.1.7
31987	app-shells/starship-1.3.0
31973	dev-lang/starlark-rust-0.7.0
31617	app-admin/terraform-0.14.10
31342	dev-texlive/texlive-metapost-2021
30833	www-client/firefox-91.10.0
30639	net-misc/calicoctl-3.1.3
30539	dev-texlive/texlive-langgerman-2021
30402	dev-texlive/texlive-langcyrillic-2021
30277	www-client/firefox-bin-91.10.0
29717	x11-terms/alacritty-0.9.0-r1
29366	app-metrics/unbound-telemetry-0.1.0_pre20210321-r1
29212	net-misc/felix-3.1.3
28881	dev-util/git-delta-0.12.1
28636	dev-texlive/texlive-games-2021
28586	dev-util/wachy-0.1.0_alpha6
28365	dev-texlive/texlive-langfrench-2021
28171	dev-texlive/texlive-xetex-2021-r1
26710	games-strategy/ja2-stracciatella-0.17.0
26693	net-misc/asterisk-core-sounds-1.6.1-r1
25963	sys-process/bottom-0.6.6
25523	sys-apps/bat-0.20.0
25475	sys-apps/lsd-0.20.1
25475	app-misc/skim-0.9.3
25192	dev-util/bingrep-0.10.0
24920	net-misc/calico-cni-plugin-3.1.3
24524	app-admin/cli53-0.8.18
23819	dev-texlive/texlive-context-2021
23646	dev-util/cargo-ebuild-0.4.0
22550	app-metrics/dnsmasq_exporter-0.2.0-r1
```
</details>

That said, I feel like the results are off by a fair bit. I suppose e.g. some Go/Rust packages are going to share the same distfiles over multiple versions, so they would be counted differently, and the effective sizes would change unpredictably over version bumps/removals.

For this reason, I'm thinking it might be more predictable to count sizes of Manifest entries per ebuild rather than look at the total file size (or perhaps even set a limit on Manifest lines per ebuild rather than size). WDYT?

File: src/pkgcheck/checks/pkgdir.py
Comment (arthurzam) [reply src/pkgcheck/checks/pkgdir.py:317]: I agree. Easier to read is more important in this place.

File: src/pkgcheck/checks/pkgdir.py
Comment (arthurzam) [inline src/pkgcheck/checks/pkgdir.py:277]: Can you please use `pathlib.Path` here - I plan to migrate from `pjoin` to `pathlib` for better readability and safety.

File: src/pkgcheck/checks/pkgdir.py
Comment (arthurzam) [inline src/pkgcheck/checks/pkgdir.py:287]: Please move this check into separate function inside this `Check` class. This is easier to understand, and will make the short-circuit logic easier.
Then when you call the function, use something like `yield from self.<func-name>(...)`

File: tests/checks/test_pkgdir.py
Comment (arthurzam) [inline tests/checks/test_pkgdir.py:299]: I'm not that sure about this code. My knowledge about filesystems is lacking in this regard, and I'm unsure if this will always work on all filesystems.
Maybe add small `assert` in this `write_manifest`that check the size of this file matches (and then add good assert message that the test setup is wrong on this FS and don't blame the package).

File: tests/checks/test_pkgdir.py
Comment (arthurzam) [inline tests/checks/test_pkgdir.py:309]: Maybe it would be better to use relative size here (I know 512 is really small number, but just in case we decide to change units or something like this)? I'm not sure - so your call 😄 

```suggestion
        self.write_manifest(pkg, pkgdir.ManifestSizeViolation.limit_per_version / 2)
```

**Comment by arthurzam** [comment]
> Ok, I've added a check for Manifest size as well. Put a random figure of 100 KiB per package version for now.

Fine, before merge we need to check/select the desired limits.

> I've made the limit dependent on number of unique package versions, so that e.g. version bumps don't cause packages to suddenly just over limit, and then cleanups put them below again. So e.g. a package with 3 versions get 300 KiB limit.
> 
> 
> That said, I feel like the results are off by a fair bit. I suppose e.g. some Go/Rust packages are going to share the same distfiles over multiple versions, so they would be counted differently, and the effective sizes would change unpredictably over version bumps/removals.
> 
> For this reason, I'm thinking it might be more predictable to count sizes of Manifest entries per ebuild rather than look at the total file size (or perhaps even set a limit on Manifest lines per ebuild rather than size). WDYT?

I also prefer count of lines. One day, we might change the used hashes in Manifest, which will break the size count for a lot of package. Lines count is safer, and easier to count across ebuilds (count the number of different items in pkg.SRC_URI).

**Comment by mgorny** [comment]
Yes, that's a good argument too.

I've experimented with distfile counts now, and the winners are…

<details><summary>Ebuilds in ::gentoo with highest distfile count</summary>

```
3514 ebuild src: dev-texlive/texlive-latexextra-2021
2044 ebuild src: app-metrics/prometheus-2.31.1
1795 ebuild src: app-admin/filebeat-7.17.1
1732 ebuild src: app-admin/filebeat-7.16.2
1551 ebuild src: app-misc/spire-1.2.0
1430 ebuild src: app-misc/spire-1.1.0
1200 ebuild src: app-containers/distrobuilder-1.2
1183 ebuild src: net-misc/rclone-1.57.0
1181 ebuild src: app-admin/terraform-1.1.5
1181 ebuild src: app-admin/terraform-1.1.2
1174 ebuild src: net-analyzer/telegraf-1.18.3
1055 ebuild src: app-admin/terraform-1.0.9
1055 ebuild src: app-admin/terraform-1.0.8
1055 ebuild src: app-admin/terraform-1.0.7
1055 ebuild src: app-admin/terraform-1.0.6
1055 ebuild src: app-admin/terraform-1.0.10
1049 ebuild src: app-admin/terraform-1.0.5
1049 ebuild src: app-admin/terraform-1.0.4
1049 ebuild src: app-admin/terraform-1.0.3
1034 ebuild src: app-admin/terraform-1.0.1
1034 ebuild src: app-admin/terraform-0.15.5
1018 ebuild src: net-analyzer/telegraf-1.18.2
1018 ebuild src: net-analyzer/telegraf-1.18.1
1018 ebuild src: net-analyzer/telegraf-1.18.0
920  ebuild src: app-admin/terraform-0.14.10
895  ebuild src: net-analyzer/telegraf-1.15.2
881  ebuild src: net-analyzer/telegraf-1.14.5
879  ebuild src: net-fs/minio-2021.04.18.19.26.29-r2
784  ebuild src: net-misc/rclone-1.53.3
784  ebuild src: net-fs/mc-2021.03.23.05.46.11
744  ebuild src: app-backup/restic-0.12.1
742  ebuild src: app-backup/restic-0.13.1
742  ebuild src: app-backup/restic-0.13.0
730  ebuild src: games-board/tablebase-syzygy-0_pre20131213
727  ebuild src: dev-texlive/texlive-fontsextra-2021
591  ebuild src: dev-texlive/texlive-publishers-2021
591  ebuild src: app-backup/restic-0.11.0
578  ebuild src: app-metrics/node_exporter-1.3.1
572  ebuild src: app-metrics/node_exporter-1.2.2
572  ebuild src: app-metrics/node_exporter-1.2.0
562  ebuild src: app-metrics/node_exporter-1.1.2-r1
540  ebuild src: dev-texlive/texlive-mathscience-2021
537  ebuild src: app-shells/nushell-0.63.1
531  ebuild src: app-admin/gopass-1.14.1
525  ebuild src: dev-embedded/arduino-builder-1.5.4-r1
522  ebuild src: app-shells/nushell-0.43.0
519  ebuild src: app-shells/nushell-0.62.0-r1
500  ebuild src: app-admin/gopass-1.13.0
489  ebuild src: app-shells/nushell-0.61.0
488  ebuild src: app-admin/gopass-1.12.7
484  ebuild src: app-admin/gopass-1.12.6
481  ebuild src: dev-texlive/texlive-pictures-2021
479  ebuild src: app-shells/nushell-0.60.0
455  ebuild src: dev-db/influxdb-1.8.3
449  ebuild src: dev-db/influxdb-1.8.2
449  ebuild src: dev-db/influxdb-1.8.1
449  ebuild src: dev-db/influxdb-1.8.0
449  ebuild src: app-metrics/blackbox_exporter-0.19.0
435  ebuild src: dev-db/influxdb-1.8.0-r1
433  ebuild src: app-metrics/postgres_exporter-0.9.0
433  ebuild src: app-metrics/memcached_exporter-0.9.0
431  ebuild src: net-analyzer/chronograf-1.8.5
404  ebuild src: app-emulation/ruffle-0_p20220601
404  ebuild src: app-emulation/ruffle-0_p20220516
383  ebuild src: net-p2p/arti-0.4.0
383  ebuild src: dev-util/sccache-0.2.15
382  ebuild src: net-p2p/arti-0.3.0
380  ebuild src: app-crypt/sequoia-sq-0.26.0
334  ebuild src: dev-go/delve-1.7.2
327  ebuild src: dev-texlive/texlive-bibtexextra-2021
321  ebuild src: dev-util/maturin-0.12.19
319  ebuild src: dev-util/maturin-0.12.18
314  ebuild src: dev-util/maturin-0.12.17
294  ebuild src: www-apps/nextcloud-notify_push-0.2.2
290  ebuild src: games-board/tablebase-syzygy-0_pre20131213-r100
288  ebuild src: www-apps/nextcloud-notify_push-0.2.1
288  ebuild src: www-apps/nextcloud-notify_push-0.2.0
283  ebuild src: www-apps/nextcloud-notify_push-0.1.7
283  ebuild src: dev-util/cargo-audit-0.15.2
282  ebuild src: www-apps/nextcloud-notify_push-0.2.4
280  ebuild src: dev-vcs/reposurgeon-4.27
279  ebuild src: app-shells/starship-1.6.3
277  ebuild src: www-apps/nextcloud-notify_push-0.3.0
266  ebuild src: dev-util/cargo-audit-0.15.0
263  ebuild src: dev-util/rustup-1.24.3
260  ebuild src: sys-apps/uutils-0.0.13
257  ebuild src: media-video/rav1e-0.4.1-r1
253  ebuild src: dev-texlive/texlive-pstricks-2021
241  ebuild src: app-shells/starship-1.4.2
240  ebuild src: app-shells/starship-1.3.0
234  ebuild src: sys-apps/uutils-0.0.12
234  ebuild src: app-shells/starship-1.5.4
230  ebuild src: app-admin/gopass-jsonapi-1.11.1
221  ebuild src: dev-texlive/texlive-plaingeneric-2021
216  ebuild src: app-text/mdbook-0.4.18
210  ebuild src: app-misc/broot-1.13.1
210  ebuild src: app-misc/broot-1.12.0
210  ebuild src: app-misc/broot-1.11.1
203  ebuild src: app-admin/gopass-hibp-1.12.0
202  ebuild src: x11-terms/alacritty-0.10.1
195  ebuild src: x11-terms/alacritty-0.9.0-r1
195  ebuild src: dev-go/golicense-0.2.0
193  ebuild src: app-text/texlive-core-2021-r1
190  ebuild src: dev-texlive/texlive-latexrecommended-2021
185  ebuild src: app-admin/git-credential-gopass-1.12.0
184  ebuild src: app-i18n/yaskkserv2-0.1.1-r1
183  ebuild src: app-office/libreoffice-l10n-7.3.4.2
183  ebuild src: app-office/libreoffice-l10n-7.3.3.2
183  ebuild src: app-office/libreoffice-l10n-7.2.6.2
179  ebuild src: sys-process/below-0.5.0
172  ebuild src: media-video/rav1e-0.5.1
172  ebuild src: dev-util/git-delta-0.13.0
170  ebuild src: dev-util/git-delta-0.12.1
165  ebuild src: sys-process/bottom-0.6.8
165  ebuild src: sys-process/bottom-0.6.6
165  ebuild src: dev-lang/starlark-rust-0.8.0
163  ebuild src: net-misc/websocat-1.9.0
163  ebuild src: dev-util/wasmer-0.11.0
161  ebuild src: sys-apps/ripgrep-all-0.9.6-r1
159  ebuild src: net-misc/websocat-1.6.0-r2
158  ebuild src: dev-texlive/texlive-latex-2021
150  ebuild src: app-crypt/sequoia-sqv-1.1.0
149  ebuild src: dev-util/cargo-ebuild-0.4.0
148  ebuild src: net-libs/quiche-0.14.0
147  ebuild src: net-libs/quiche-0.13.0
146  ebuild src: sys-apps/bat-0.21.0
145  ebuild src: net-vpn/vpncloud-2.3.0
144  ebuild src: sys-apps/bat-0.20.0
141  ebuild src: dev-texlive/texlive-langeuropean-2021
140  ebuild src: dev-util/cargo-c-0.9.6
140  ebuild src: dev-texlive/texlive-luatex-2021
140  ebuild src: dev-lang/starlark-rust-0.7.0
139  ebuild src: dev-util/cargo-c-0.9.5
138  ebuild src: dev-util/cargo-c-0.9.9
135  ebuild src: dev-util/cargo-c-0.9.7
134  ebuild src: dev-texlive/texlive-humanities-2021
134  ebuild src: app-admin/gopass-summon-provider-1.12.0
133  ebuild src: dev-util/cargo-c-0.6.2-r1
132  ebuild src: net-libs/quiche-0.12.0
132  ebuild src: net-libs/quiche-0.11.0
132  ebuild src: gui-apps/tuigreet-0.7.2
131  ebuild src: net-libs/quiche-0.9.0-r1
131  ebuild src: net-libs/quiche-0.10.0
131  ebuild src: dev-texlive/texlive-langother-2021
129  ebuild src: dev-util/cargo-ebuild-0.5.0
129  ebuild src: dev-go/go-licenses-0_pre20200511
129  ebuild src: app-metrics/fritzbox_smarthome_exporter-0.0.8
128  ebuild src: dev-vcs/mercurial-6.1.3-r1
128  ebuild src: dev-util/cargo-ebuild-0.5.1
127  ebuild src: dev-texlive/texlive-langenglish-2021
```
</details>

**Comment by mgorny** [comment]
So I was thinking about it more and perhaps the simplest solution (i.e. plain Manifest sizes) is the best one. Why?

1. Because it's easy to understand and check for devs, so it doesn't come off as unpredictable.
2. Because it matches the end goal, i.e. reducing the space actually consumed on end-user systems.
3. Because it gives the developers the ability to choose a compromise between keeping multiple old versions and reducing distfile list length.

In other words, seeing "foo suddenly violates the rule after version bump" not as a limitation of the check itself but as an indication that the package is problematic and needs to reduce the distfile length to account for these bumps.

<details><summary>Biggest Manifests in ::gentoo</summary>

```
1095871 ./dev-texlive/texlive-latexextra/Manifest
699709  ./app-metrics/prometheus/Manifest
622870  ./app-admin/filebeat/Manifest
539935  ./app-misc/spire/Manifest
505882  ./app-admin/terraform/Manifest
496032  ./net-analyzer/telegraf/Manifest
486110  ./net-misc/rclone/Manifest
413286  ./app-containers/distrobuilder/Manifest
314204  ./app-shells/nushell/Manifest
306953  ./app-metrics/node_exporter/Manifest
305354  ./games-board/tablebase-syzygy/Manifest
305165  ./app-backup/restic/Manifest
298904  ./net-fs/minio/Manifest
266770  ./net-fs/mc/Manifest
226363  ./dev-texlive/texlive-fontsextra/Manifest
207216  ./app-admin/gopass/Manifest
192504  ./www-apps/nextcloud-notify_push/Manifest
187855  ./app-office/libreoffice-l10n/Manifest
184518  ./dev-texlive/texlive-publishers/Manifest
184460  ./dev-embedded/arduino-builder/Manifest
168333  ./dev-texlive/texlive-mathscience/Manifest
158488  ./dev-db/influxdb/Manifest
153441  ./app-metrics/blackbox_exporter/Manifest
149972  ./dev-texlive/texlive-pictures/Manifest
147704  ./app-metrics/memcached_exporter/Manifest
147618  ./app-metrics/postgres_exporter/Manifest
146924  ./net-analyzer/chronograf/Manifest
128060  ./app-emulation/ruffle/Manifest
127948  ./app-shells/starship/Manifest
126762  ./net-p2p/arti/Manifest
117535  ./dev-util/sccache/Manifest
116566  ./app-crypt/sequoia-sq/Manifest
114174  ./dev-go/delve/Manifest
112293  ./dev-util/cargo-c/Manifest
110962  ./media-video/rav1e/Manifest
110627  ./app-shells/bash/Manifest
104912  ./dev-util/cargo-audit/Manifest
104896  ./dev-util/maturin/Manifest
102704  ./dev-texlive/texlive-bibtexextra/Manifest
96534   ./sys-apps/uutils/Manifest
95816   ./dev-vcs/reposurgeon/Manifest
92499   ./www-client/firefox/Manifest
89151   ./x11-terms/alacritty/Manifest
83851   ./net-libs/quiche/Manifest
82170   ./app-text/tessdata_legacy/Manifest
80770   ./dev-util/rustup/Manifest
80088   ./app-text/tessdata_best/Manifest
79958   ./app-text/tessdata_fast/Manifest
78972   ./dev-texlive/texlive-pstricks/Manifest
77948   ./app-admin/gopass-jsonapi/Manifest
76507   ./net-misc/websocat/Manifest
70938   ./dev-util/cargo-ebuild/Manifest
68941   ./app-admin/gopass-hibp/Manifest
68414   ./dev-texlive/texlive-plaingeneric/Manifest
67037   ./dev-go/golicense/Manifest
66250   ./app-text/mdbook/Manifest
65299   ./app-misc/broot/Manifest
63946   ./dev-lang/starlark-rust/Manifest
62805   ./app-admin/git-credential-gopass/Manifest
60554   ./www-client/firefox-bin/Manifest
60053   ./app-text/texlive-core/Manifest
59128   ./dev-texlive/texlive-latexrecommended/Manifest
58733   ./app-metrics/unbound-telemetry/Manifest
57762   ./dev-util/git-delta/Manifest
56431   ./app-i18n/yaskkserv2/Manifest
54971   ./sys-process/below/Manifest
51926   ./sys-process/bottom/Manifest
51046   ./sys-apps/bat/Manifest
50951   ./sys-apps/lsd/Manifest
50009   ./dev-util/wasmer/Manifest
49453   ./sys-apps/ripgrep-all/Manifest
49254   ./dev-texlive/texlive-latex/Manifest
47017   ./dev-vcs/mercurial/Manifest
46489   ./dev-ros/test_rosbag/Manifest
45983   ./app-crypt/sequoia-sqv/Manifest
45476   ./app-admin/gopass-summon-provider/Manifest
44497   ./net-vpn/vpncloud/Manifest
44434   ./dev-texlive/texlive-langeuropean/Manifest
43888   ./net-news/newsboat/Manifest
43791   ./dev-go/go-licenses/Manifest
43783   ./app-metrics/fritzbox_smarthome_exporter/Manifest
43675   ./dev-texlive/texlive-luatex/Manifest
43617   ./dev-util/drone-cli/Manifest
41890   ./dev-go/go-license-detector/Manifest
41611   ./dev-texlive/texlive-humanities/Manifest
41060   ./dev-texlive/texlive-langother/Manifest
40860   ./app-misc/rpick/Manifest


```
</details>

**Comment by mgorny** [comment]
Ok, let's defer Manifest size checks into a separate PR and just focus on total `files/` size here.

File: src/pkgcheck/checks/pkgdir.py
Comment (mgorny) [reply src/pkgcheck/checks/pkgdir.py]: Done.

File: src/pkgcheck/checks/pkgdir.py
Comment (arthurzam) [inline src/pkgcheck/checks/pkgdir.py:142]: let's add comment of units here for the size:
```suggestion
    limit = 51200 # bytes = 50KiB
```

File: src/pkgcheck/checks/pkgdir.py
Comment (arthurzam) [inline src/pkgcheck/checks/pkgdir.py:126]: let's add comment of units here for the size:
```suggestion
    limit = 20480 # bytes = 20KiB
```

**Comment by mgorny** [comment]
Done.

**Review by arthurzam — APPROVED** [review]

**Pull request closed without merging**
